lib/AST/DeclHead.cpp

```diff
@@ -425,6 +427,195 @@ TemplateParameterList::Parameters(void) const noexcept {
+}
+
+std::optional<bool> CXXRecordDecl::HasOwnVirtualFunctionTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasOwnVFPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasExtendableVirtualFunctionTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasExtendableVFPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasVirtualBaseTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasVBPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasOwnVirtualBaseTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasOwnVBPtr();
+}
+
+std::optional<uint64_t> FieldDecl::OffsetInBits(void) const noexcept {
+  const clang::RecordDecl *record = u.FieldDecl->getParent();
+  if (!record) {
+    return std::nullopt;
+  }
+
+  record = record->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  const clang::Type *type = record->getTypeForDecl();
+  clang::NamedDecl *nd = nullptr;
+  if (!type || type->isIncompleteType(&nd)) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.getFieldOffset(u.FieldDecl->getFieldIndex());
+}
+
+// Clang deduplicates function types, and stores the ellipsis in those types.
+// that results in the wrong ellipsis location.
+::pasta::Token FunctionDecl::EllipsisToken(void) const {
+  auto &self = *const_cast<clang::FunctionDecl *>(u.FunctionDecl);
+  if (self.getEllipsisLoc().isValid()) {
+    auto [begin_tok, end_tok] = ast->DeclBounds(&self);
+    if (auto it = ast->func_proto.find(&self); it != ast->func_proto.end()) {
+      assert(begin_tok < it->second.ellipsis);
+      assert(it->second.ellipsis < end_tok);
+      (void) begin_tok;
+      return ::pasta::Token(ast, it->second.ellipsis);
+    }
+  }
+  return ::pasta::Token(ast);
+}
+
+::pasta::TokenRange FunctionDecl::ParametersTokens(void) const {
+  auto &self = *const_cast<clang::FunctionDecl *>(u.FunctionDecl);
+  (void) ast->DeclBounds(&self);
```

Comment:
Refactor/Defend`DeclBounds` this is highly suspicious use case and gives a feeling it is not designed properly.

Also `std::ignore`.